services/tasks/src/test/java/org/hyperledger/besu/services/tasks/InMemoryTasksPriorityQueuesTest.java

```diff
@@ -0,0 +1,175 @@
+    }
+
+    assertThat(items)
+        .containsExactly(
+            item(99, Integer.MAX_VALUE),
+            item(5, 1),
+            item(2, 10),
+            item(2, 30),
+            item(1, 1),
+            item(1, 20));
+  }
+
+  @Test
+  public void shouldNotInsertItemsToClosedQueue() {
+    InMemoryTasksPriorityQueues<Item> queue = new InMemoryTasksPriorityQueues<>();
+
+    queue.add(item(1, 1));
+
+    queue.close();
+
+    final Item item = item(2, 2);
+    assertThatThrownBy(() -> queue.add(item)).isInstanceOf(IllegalStateException.class);
+  }
+
+  @Test
+  public void shouldContainTaskUntilFinished() {
+    InMemoryTasksPriorityQueues<Item> queue = new InMemoryTasksPriorityQueues<>();
+    queue.add(item(1, 1));
+
+    final Item item = item(2, 3);
+    queue.add(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    final Task<Item> removed = queue.remove();
+    assertThat(removed.getData()).isEqualTo(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    removed.markCompleted();
+
+    assertThat(queue.contains(item)).isFalse();
+  }
+
+  @Test
+  public void shouldPutFailedItemBackIntoQueue() {
+    InMemoryTasksPriorityQueues<Item> queue = new InMemoryTasksPriorityQueues<>();
+    queue.add(item(1, 1));
+
+    final Item item = item(2, 3);
+    queue.add(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    Task<Item> removed = queue.remove();
+    assertThat(removed.getData()).isEqualTo(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    removed.markFailed();
+
+    assertThat(queue.contains(item)).isTrue();
+
+    removed = queue.remove();
+    assertThat(removed.getData()).isEqualTo(item);
+  }
+
+  @Test
+  public void shouldNotPutFailedItemBackIntoIfItWasCompletedAlreadyQueue() {
+    InMemoryTasksPriorityQueues<Item> queue = new InMemoryTasksPriorityQueues<>();
+    queue.add(item(1, 1));
+
+    final Item item = item(2, 3);
+    queue.add(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    Task<Item> removed = queue.remove();
+    assertThat(removed.getData()).isEqualTo(item);
+
+    assertThat(queue.contains(item)).isTrue();
+
+    removed.markCompleted();
+    assertThat(queue.contains(item)).isFalse();
+
+    removed.markFailed();
+    assertThat(queue.contains(item)).isFalse();
+
+    removed = queue.remove();
+    assertThat(removed.getData()).isNotEqualTo(item);
+  }
+
+  private Item item(final int depth, final int priority) {
+    return new Item(depth, priority);
+  }
+
+  static class Item implements TasksPriorityProvider {
+    private final int depth;
+    private final long priority;
+
+    public Item(final int depth, final long priority) {
+      this.depth = depth;
+      this.priority = priority;
+    }
+
+    @Override
+    public long getPriority() {
+      return priority;
+    }
+
+    @Override
+    public int getDepth() {
+      return depth;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```

Comment:
this class implements equals and hashCode. Is it expected all impls of `TasksPriorityProvider` implement those methods? I don't see them implemented for all of them.

Reply:
I used the equals/hashcode/toString only when creating the testcases. The Interface is expecting to see the same instance of objects at the moment (as you noted with the == used in the contains method)